src/duckdb.mjs

```diff
@@ -1,361 +1,234 @@
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+//    this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+//    this list of conditions and the following disclaimer in the documentation
+//    and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+// TODO Allow this to be overridden using the Library’s resolver.
+const cdn = "https://cdn.observableusercontent.com/npm/";
+
+export class DuckDBClient {
+  constructor(db) {
+    Object.defineProperties(this, {
+      _db: {value: db}
+    });
   }
 
-  // Adapted from: https://observablehq.com/@cmudig/duckdb-client
-  // Follows the DatabaseClient specification: https://observablehq.com/@observablehq/database-client-specification
-  class DuckDBClient {
-    constructor(_db) {
-      this._db = _db;
-      this._counter = 0;
+  async queryStream(query, params) {
+    const connection = await this._db.connect();
+    let reader, schema, batch;
+    try {
+      reader = await connection.send(query, params);
+      batch = await reader.next();
+      if (batch.done) throw new Error("missing first batch");
+      schema = batch.value.schema;
+    } catch (error) {
+      await connection.close();
+      throw error;
     }
-
-    async queryStream(query, params) {
-      const conn = await this.connection();
-      let result;
-
-      if (params) {
-        const stmt = await conn.prepare(query);
-        result = await stmt.query(...params);
-      } else {
-        result = await conn.query(query);
-      }
-      // Populate the schema of the results
-      const schema = result.schema.fields.map(({
-        name,
-        type
-      }) => ({
+    return {
+      schema: schema.fields.map(({name, type}) => ({
         name,
         type: getType(String(type)),
```

Comment:
This is maybe more brittle than directly checking against Arrow types but probably okay for now. Can improve later.

Reply:
Yes, I can use the Apache Arrow type identifiers when looking at the schema for an Apache Arrow Table, but the result of a DESCRIBE doesn’t give me the numeric identifiers — I only get the string column_type. I want to use the same code for describeColumns, too. It’d be nice if DuckDB returned the numeric type identifier in addition to the string.

Reply:
Added in 990c240. And I was able to simplify the switch for DuckDB types since now we only need to handle the canonical types returned by DESCRIBE rather than the aliases. (At least, I hope that’s the case… it appears to be in my local testing.)